[UnifiedPDF] Context menu should have "Next Page" and "Previous Page" navigation options.

sammygill opened the pull request:
#### 799678c6dd8cd120146f7acb293ba006761b569d
<pre>
[UnifiedPDF] Context menu should have &quot;Next Page&quot; and &quot;Previous Page&quot; navigation options.
<a href="https://bugs.webkit.org/show_bug.cgi?id=269347">https://bugs.webkit.org/show_bug.cgi?id=269347</a>
<a href="https://rdar.apple.com/122931085">rdar://122931085</a>

Reviewed by Tim Horton.

Adds the next and previous page items to the context menu. The &quot;Next
Page,&quot; option is only able to be selected if the current page that is
in the view is not the last page. Likewise, the &quot;Previous Page,&quot; option
is only selectable if the current page is not the first page. Regardless
both of the options are shown but are disabled in those scenarios.

In order to determine which page we are on we take the very center of
the view and see which page that point is contained in. This code
currently works only for single page display modes so a follow up will
be needed to get it working for two page display modes.

For scrolling between the pages we take the layout bounds, convert them
to the contents space, subtract out the page margins, and then
perform a scroll without animation to that point. Modified
scrollToPointInPage to use this new helper function that subtracts out
the page margins instead of the document margins like it was before and
which I don&apos;t think was correct.

Also rearranged the context menu slightly so that the ordering of the
items matches the current order displayed with the legacy plugin.

* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::contextMenuItemPDFActualSize):
(WebCore::contextMenuItemPDFAutoSize):
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::indexForCurrentPageInView const):
(WebKit::UnifiedPDFPlugin::convertFromPageToContents const):
(WebKit::UnifiedPDFPlugin::convertFromDocumentToContents const):
(WebKit::UnifiedPDFPlugin::offsetContentsSpacePointByPageMargins const):
(WebKit::UnifiedPDFPlugin::scrollToPointInPage):
(WebKit::UnifiedPDFPlugin::scrollToPage):
(WebKit::UnifiedPDFPlugin::toContextMenuItemTag const):
(WebKit::UnifiedPDFPlugin::createContextMenu const):
(WebKit::UnifiedPDFPlugin::titleForContextMenuItemTag const):
(WebKit::UnifiedPDFPlugin::contextMenuItem const):
(WebKit::UnifiedPDFPlugin::navigationContextMenuItems const):
(WebKit::UnifiedPDFPlugin::performContextMenuAction):

Canonical link: <a href="https://commits.webkit.org/274737@main">https://commits.webkit.org/274737@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a571d9d681b59137a4f1f8b7e616f6dad499ed44

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39803 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18814 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42178 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42347 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35705 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42110 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21754 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16144 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33187 "Found 1 new test failure: http/tests/scroll-to-text-fragment/no-scroll-after-stylesheet-load.html (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40377 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15887 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34419 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13720 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13774 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35374 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43626 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36177 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35708 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39479 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14618 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12025 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37764 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16237 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8946 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16285 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15881 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1225,6 +1225,12 @@ static String mutationObserverNotificationString()
+std::optional<PDFDocumentLayout::PageIndex> UnifiedPDFPlugin::indexForCurrentPageInView() const
+{
+    auto centerInDocumentSpace = convertFromPluginToDocument(flooredIntPoint(size() / 2));
+    return pageIndexForDocumentPoint(centerInDocumentSpace);
```
Does this find the nearest page, or does it just fail sometimes?

sammygill replied:
Yeah so I couldn't actually get something to reproduce locally but after taking a closer look at the code I'm pretty sure it can fail. I think the document layout bounds do not include margins which is the problem. 

Also it turns out this does not work well for two page display modes so I'm going to fix both in a follow up patch. Made a note in the commit message stating two page display has issues

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1225,6 +1225,12 @@ static String mutationObserverNotificationString()
 
+std::optional<PDFDocumentLayout::PageIndex> UnifiedPDFPlugin::indexForCurrentPageInView() const
+{
+    auto centerInDocumentSpace = convertFromPluginToDocument(flooredIntPoint(size() / 2));
```
I didn't expect us to find the current page index like this

aprotyas replied:
I was thinking you'd go off of the context menu event position

aprotyas replied:
Wouldn't that be more reliable?

sammygill replied:
That is what I was initially planning on doing but our current behavior seems to go based off the page that is in the center of the view regardless of where the event occurred

aprotyas replied:
Ah, I see what you mean.

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1662,11 +1682,17 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
 
-    menuItems.appendVector(scaleContextMenuItems());
+    auto currentPageIndex = indexForCurrentPageInView();
+    menuItems.append(contextMenuItem(ContextMenuItemTag::NextPage, currentPageIndex != m_documentLayout.pageCount() - 1));
```
Maybe make another group? navigationContextMenuItems?

sammygill replied:
Done!

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1770,6 +1798,15 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
+{
+    auto currentPageIndex = indexForCurrentPageInView();
+    return {
+    contextMenuItem(ContextMenuItemTag::NextPage, currentPageIndex != m_documentLayout.pageCount() - 1),
```
Indentation looks off here.

sammygill committed:
[UnifiedPDF] Context menu should have "Next Page" and "Previous Page" navigation options.
https://bugs.webkit.org/show_bug.cgi?id=269347
rdar://122931085

Reviewed by Tim Horton.

Adds the next and previous page items to the context menu. The "Next
Page," option is only able to be selected if the current page that is
in the view is not the last page. Likewise, the "Previous Page," option
is only selectable if the current page is not the first page. Regardless
both of the options are shown but are disabled in those scenarios.

In order to determine which page we are on we take the very center of
the view and see which page that point is contained in. This code
currently works only for single page display modes so a follow up will
be needed to get it working for two page display modes.

For scrolling between the pages we take the layout bounds, convert them
to the contents space, subtract out the page margins, and then
perform a scroll without animation to that point. Modified
scrollToPointInPage to use this new helper function that subtracts out
the page margins instead of the document margins like it was before and
which I don't think was correct.

Also rearranged the context menu slightly so that the ordering of the
items matches the current order displayed with the legacy plugin.

* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::contextMenuItemPDFActualSize):
(WebCore::contextMenuItemPDFAutoSize):
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::indexForCurrentPageInView const):
(WebKit::UnifiedPDFPlugin::convertFromPageToContents const):
(WebKit::UnifiedPDFPlugin::convertFromDocumentToContents const):
(WebKit::UnifiedPDFPlugin::offsetContentsSpacePointByPageMargins const):
(WebKit::UnifiedPDFPlugin::scrollToPointInPage):
(WebKit::UnifiedPDFPlugin::scrollToPage):
(WebKit::UnifiedPDFPlugin::toContextMenuItemTag const):
(WebKit::UnifiedPDFPlugin::createContextMenu const):
(WebKit::UnifiedPDFPlugin::titleForContextMenuItemTag const):
(WebKit::UnifiedPDFPlugin::contextMenuItem const):
(WebKit::UnifiedPDFPlugin::navigationContextMenuItems const):
(WebKit::UnifiedPDFPlugin::performContextMenuAction):

Canonical link: https://commits.webkit.org/274737@main
